62.CAD/Hatch.cpp

```diff
@@ -0,0 +1,782 @@
+    
+	// Intended to mitigate issues with NaNs and precision by falling back to using simpler functions when the higher roots are small enough
+	static std::array<double, 4> solveQuarticRoots(double a, double b, double c, double d, double e, double t_start, double t_end)
+	{
+		std::array<double, 4> t = { -1.0, -1.0, -1.0, -1.0 }; // only two candidates in range, ever
+
+		const double quadCoeffMag = std::max(std::abs(d), std::abs(e));
+		const double cubCoeffMag = std::max(std::abs(c), quadCoeffMag);
+		if (std::abs(a) > std::max(std::abs(b), cubCoeffMag) * QUARTIC_THRESHHOLD)
+		{
+			auto res = equations::Quartic<double>::construct(a, b, c, d, e).computeRoots();
+			memcpy(&t[0], &res.x, sizeof(double) * 4);
+		}
+		else if (abs(b) > quadCoeffMag)
+		{
+			auto res = equations::Cubic<double>::construct(b, c, d, e).computeRoots();
+			memcpy(&t[0], &res.x, sizeof(double) * 3);
+		}
+		else
+		{
+			auto res = equations::Quadratic<double>::construct(c, d, e).computeRoots();
+			memcpy(&t[0], &res.x, sizeof(double) * 2);
+		}
+
+		if (!(t[0] != t[1]))
+			t[0] = t[0] != t_start ? t_start : t_end;
+
+		//// TODO: check that this clamp works with t[i] as NaN
+		//for (auto i = 0; i < 2; i++)
+		//	t[i] = nbl::core::clamp(t[i], t_start, t_end);
+		//
+		//// fix up a fuckup where both roots are NaN or were on the same side of the valid integral
+		//
+		//// TODO: do some Halley or Householder method steps on t
+		////while ()
+		////{
+		////}
+
+		// neither t still not in range, your beziers don't intersect
+		return t;
+	}
+
+	// (only works for monotonic curves)
+	static float64_t2 getCurveRoot(double p0, double p1, double p2)
+	{
+		// https://people.csail.mit.edu/bkph/articles/Quadratics.pdf
+		double a = p0 - 2.0 * p1 + p2;
+		double b = 2.0 * (p1 - p0);
+		double c = p0;
+
+		// Solve linear equation
+		if (a == 0.0) return float64_t2(-c / b);
+
+		double det = sqrt(b * b - 4 * a * c);
+		if (b >= 0.0) {
+			return float64_t2(
+				(- b - det) / (2 * a),
+				(2 * c) / (-b - det)
+			);
+		}
+		else
+		{
+			return float64_t2(
+				(2 * c) / (-b + det),
+				(-b + det) / (2 * a)
+			);
+		}
+		//double det = b * b - 4 * a * c;
+		//double rcp = 0.5 / a;
+		//
+		//double detSqrt = sqrt(det) * rcp;
+		//double brcp = b * rcp;
+		//
+		//// Solve linear equation
+		//if (isinf(brcp)) return float64_t2(-c / b);
+		//
+		//return float64_t2(-detSqrt, detSqrt) - brcp;
+	}
+};
+
+Hatch::QuadraticBezier Hatch::Segment::splitCurveRange(QuadraticBezier curve, double left, double right) const
+{
+    return curve.splitCurveTakeLower(right).splitCurveTakeUpper(left / right);
+}
+
+Hatch::QuadraticBezier Hatch::Segment::getSplitCurve() const
+{
+    return splitCurveRange(*originalBezier, t_start, t_end);
+}
+
+bool Hatch::Segment::isStraightLineConstantMajor() const
+{
+	auto major = (uint)SelectedMajorAxis;
+	return originalBezier->p[0][major] == originalBezier->p[1][major] &&
+		originalBezier->p[0][major] == originalBezier->p[2][major];
+}
+
```

Comment:
@devshgraphicsprogramming any ideas on how to do these checks here with a certain threshold and safety?

Reply:
I'd first and foremost assert that `p0<=p1<=p2` cause otherwise the monotonicty is in question.

then you can do the check about p2 being at p0 or at a small enough epsilon (since you later store Beziers in UV space which is [0,1] fp32, I guess use 2^-24 as the threshold?)

Reply:
We're comparing three values, we should check p0, p1 and p2 major component being relatively close together to know if they form a line which is constant in major

Reply:
@devshgraphicsprogramming they wouldn't be in [0,1] yet at this point, that's only after the curve box logic

Reply:
> We're comparing three values, we should check p0, p1 and p2 major component being relatively close together to know if they form a line which is constant in major

because `assert(p0<=p1 && p1<=p2)` its enough to check `p0` and `p2` being close together

> @devshgraphicsprogramming they wouldn't be in [0,1] yet at this point, that's only after the curve box logic

Then `exp2(-24)*aabb.extent()[major]` I guess?
Btw I'm not hell bent on the -24, can be any other constant.

You probably want to make it quite big (as big as you can without introducting visual errors) because it make all other arithmetic that follows it much nicer (no divisions by small slopes or major coordinate differences).

Reply:
I don't think the aabb is constructed at this point in code without knowing other candidates and creating boxes.

So these are world space values.

And the highest threshold we can set to not have "visual" errors depends on our projection matrix (when a curve has a small slope but gets Rasterized to 1 pixel row) 
which is also weird to take into account here.

So I guess let's stick with some low world space error threshold like exp2(-20) or something

Reply:
Also the values are fp64/double here